File: MergeKLinkedLists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        if (lists.length == 0) return null; // annoying guard clause

        ListNode list = lists[0]; // starts out as the head of the first list in lists,
        // which links to the rest of the list at lists[0]

        for (int i = 1; i < lists.length; i++) {
            list = mergeTwoLists(list, lists[i]); // Assign list a new head if
            // necessary, while merging the ith list into our megalist, list
        }

        return list; // We've merged all the lists together
    }

    // Solution to a prior question
    public static ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        // Create a dummy head for the result list
        ListNode dummy = new ListNode(0);
        ListNode current = dummy;

        // Iterate through both lists
        while (list1 != null && list2 != null) {
            if (list1.val < list2.val) {
                current.next = list1;
                list1 = list1.next;
            } else {
                current.next = list2;
                list2 = list2.next;
            }
            current = current.next;
        }

        // Append the remaining elements of the non-exhausted list
        if (list1 != null) {
            current.next = list1;
        } else {
            current.next = list2;
        }

        // Return the merged list, skipping the dummy head
        return dummy.next;
    }
}

File: MergeKLinkedLists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:

        if lists is None or len(lists) == 0:
            return None

        while len(lists) > 1:
            temp_lists = []

            # Merge neighboring pairs of lists, l and r and repeat
            # Logarithmic reduction of size of number of lists, each iteration
            for i in range(0, len(lists), 2):
                l = lists[i]
                r = None # In case we're out of bounds when lists is odd length
                if i + 1 < len(lists):
                    r = lists[i + 1] # Else properly assign r
                
                # Merge l and r and save that list to merge with others
                # in future iterations
                temp_lists.append(self.merge2Lists(l, r))

            lists = temp_lists # Reassign lists, now half as long
        
        return lists[0] # Loop terminated when lists reached a length of 1
        # That is our mega-list of all elements sorted

    # This is also the solution to another problem, I priorly solved it in Java
    def merge2Lists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode() # Starting head to build our sorted list off of
        curr = dummy # curr is for iterating reference

        # Alternate between both lists, sorting them by manipulating next's
        while list1 and list2:
            # Lists are sorted so pick smaller of currently lined up values
            # And increment respective list, assign proper next, increment curr
            # to continue sorting
            if list1.val < list2.val:
                curr.next = list1
                list1 = list1.next
            else:
                curr.next = list2
                list2 = list2.next
            curr = curr.next
        
        # Ran out of list2 case
        if list1:
            curr.next = list1 # From where we left off in list1 in prior loop
        
        if list2: # Ran out of list1 case
            curr.next = list2 # From where we left off in list1 in prior loop
        
        return dummy.next # the head of the sorted (list1 + list2)
